### PR TITLE
RA-543: Patient name should not allow leading or trailing spaces

### PR DESCRIPTION
**Pull request by samuelmale**
ticket: https://issues.openmrs.org/browse/RA-543

### CODE REVIEW

**Comment by samuelmale** [comment]
@dkayiwa  I fixed broken Tests. done some channges

**Comment by djazayeri** [comment]
This is a commit to openmrs-core, but the ticket is for the Reference Application project. We'd need to have a ticket for openmrs-core (i.e. TRUNK-xyz) before merging this.

**Comment by samuelmale** [comment]
@djazayeri  I have done more changes

File: api/src/main/resources/messages.properties [hunk 1]
```diff
@@ -3291,3 +3291,8 @@ OrderFrequency.concept.shouldBeClassFrequency=Concept should be of class Frequen
 OrderFrequency.concept.shouldNotBeShared=Only one OrderFrequency should be mapped to a particular concept
 
 Module.error.cyclicDependencies=Could not start some modules because they depend on others which are not started. Please first start them or use the "Start All" button on the "Manage Modules" page: {0}
+
+GivenName.startsOrEndsWithWhitespace=Invalid GivenName, Trailing or Leading Whitespace detected
```
Comment (djazayeri) [inline api/src/main/resources/messages.properties]: the error message should be something like "Given name may not start or end with blank spaces"

**Comment by djazayeri** [comment]
Someone (whether @samuelmale or @dkayiwa) needs to check why this situation isn't already being handled by the pre-existing code that I mentioned, before we commit to merging this.

**Comment by Ben-Waters** [comment]
@djazayeri   Maybe I'm missing something but I think you're right. Running Openmrs-core version 2.2.0-SNAPSHOT  I created a new patient with both leading and trailing white space around their first middle and last names. On save, all the whitespace was removed. I then tried to edit the patient and add whitespace. On save it was all removed again.

The ticket points to a devtest2 environment which seems to be down, so I'm not able to reproduce it there. Is there a way to either start up devtest2 or find out what version it was running to try to reproduce it locally?

**Comment by dkayiwa** [comment]
@Ben-Waters did you get a chance to look at the ticket comments? I think there is one which seems to address your question. 😊

**Comment by teleivo** [comment]
@Ben-Waters or @samuelmale is this issue already handled by the pre-existing code as @djazayeri mentioned?

If not @samuelmale please resolve the conflicts in the PR.

**Comment by Ben-Waters** [comment]
When I investigated the existing code, it is fixing it as it saves to the DB but not as the user is editing and validating it.

**Comment by samuelmale** [comment]
So think we should move this to the validation layer :)

I have just fixed the merge conflicts :)

**Comment by teleivo** [comment]
@samuelmale I dont think that Validators are the right place. I would not expect a Validator to mutate something I pass into it, it should only tell me if that thing is valid and if it isnt what is invalid about it.

**Comment by samuelmale** [comment]
@teleivo  But according to my implementation in the latest commit, Validators don't mutate any values. They tell us if a value is valid or not

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 2]
```diff
@@ -129,11 +129,26 @@ public void validatePersonName(PersonName personName, Errors errors, boolean arr
...
 		}
 		// Make sure they assign a name
-		if (StringUtils.isBlank(personName.getGivenName())
+		if (StringUtils.isBlank(personName.getGivenName()) 
 		        || StringUtils.isBlank(personName.getGivenName().replaceAll("\"", ""))) {
 			errors.rejectValue(getFieldKey("givenName", arrayInd, testInd), "Patient.names.required.given.family");
 		}
 
+		// Make sure entered name value has no leading and trailing spaces
+		if (StringUtils.isNotBlank(personName.getGivenName()) && startsOrEndsWithWhitespace(personName.getGivenName())) {
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: please move `StringUtils.isNotBlank(personName.getGivenName())` into the private method so we dont have to repeat this for every field, plus if its blank it does not start or end with whitespace so this fits into the duty of the method

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 3]
```diff
@@ -164,4 +179,22 @@ private String getFieldKey(String field, boolean arrayInd, boolean testInd) {
 		return testInd ? field : arrayInd ? "names[0]." + field : "name." + field;
 	}
 	
+	/*
+	 * Checks that the given {@link PersonName} field has no trailing and leading spaces.
+	 * @param value the value name.
+	 * @return containsSpace
+	 * @should return true if given field contains trailing or leading space(s), otherwise false.
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: remove javadocs of this private method, the method has a clear name which says it all

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 4]
```diff
@@ -164,4 +179,22 @@ private String getFieldKey(String field, boolean arrayInd, boolean testInd) {
...
+	 */
+	private boolean startsOrEndsWithWhitespace(String value) {
+		boolean startsOrEndsWithWhitespace = false;
+		int lastIndex = value.length() - 1;
+		if (value.charAt(0) == ' ') {
+			startsOrEndsWithWhitespace = true;
+		} else if (Character.isWhitespace(value.charAt(lastIndex))) {
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: this also contains file separator, is this what you want? if not maybe thats better
https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#isSpaceChar(char)

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 5]
```diff
@@ -164,4 +179,22 @@ private String getFieldKey(String field, boolean arrayInd, boolean testInd) {
...
+	/*
+	 * Checks that the given {@link PersonName} field has no trailing and leading spaces.
+	 * @param value the value name.
+	 * @return containsSpace
+	 * @should return true if given field contains trailing or leading space(s), otherwise false.
+	 */
+	private boolean startsOrEndsWithWhitespace(String value) {
+		boolean startsOrEndsWithWhitespace = false;
+		int lastIndex = value.length() - 1;
+		if (value.charAt(0) == ' ') {
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: why use space here and below `Character.isWhitespace(value.charAt(lastIndex)))`?

Use the same for beginning and end.

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 6]
```diff
@@ -164,4 +179,22 @@ private String getFieldKey(String field, boolean arrayInd, boolean testInd) {
...
+			startsOrEndsWithWhitespace = true;
+		} else if (Character.isWhitespace(value.charAt(lastIndex))) {
+			startsOrEndsWithWhitespace = true; 
+		}
+		
+		return startsOrEndsWithWhitespace;
+
```
Comment (teleivo) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: remove newline

File: api/src/test/java/org/openmrs/validator/PersonNameValidatorTest.java [hunk 7]
```diff
@@ -864,4 +856,5 @@ public void validate_shouldFailValidationIfPersonNameFamilyName2HasTrailingSpace
 		
 		assertThat(errors, hasFieldErrors("familyName2"));
 	}
+
```
Comment (teleivo) [inline api/src/test/java/org/openmrs/validator/PersonNameValidatorTest.java]: remove newline

**Comment by teleivo** [comment]
@djazayeri can you please check this PR again, it has been updated.

**Comment by samuelmale** [comment]
Implemented requested changes. @teleivo

**Comment by teleivo** [comment]
@samuelmale I would like to have @djazayeri to have a look at this PR as I dont have the context.

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 8]
```diff
@@ -164,4 +180,31 @@ private String getFieldKey(String field, boolean arrayInd, boolean testInd) {
 		return testInd ? field : arrayInd ? "names[0]." + field : "name." + field;
 	}
 	
-}
+	
+	private boolean startsOrEndsWithWhitespace(String value) {
+		boolean startsOrEndsWithWhitespace = false;
+		if (value != null && StringUtils.isNotBlank(value)) {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: Isn't isNotBlank already checking for null?

Comment (samuelmale) [reply api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: Just addressed your comment :)

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 9]
```diff
@@ -164,4 +180,31 @@ private String getFieldKey(String field, boolean arrayInd, boolean testInd) {
...
+			    startsOrEndsWithWhitespace = true; 
+		    }
+		} else {
+			startsOrEndsWithWhitespace = false;
+		}
+		return startsOrEndsWithWhitespace;		
+	}
+	
+	private boolean matchNamePattern(String value, String namePattern) {
+		//True by default because this shouldNotValidateAgainstRegexForBlankNames
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: Is the above supposed to be an English sentence?

Comment (samuelmale) [reply api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: Literally meaning I get the comment off? @dkayiwa

**Comment by samuelmale** [comment]
Just removed the comment

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 10]
```diff
@@ -164,4 +180,29 @@ private String getFieldKey(String field, boolean arrayInd, boolean testInd) {
...
+		    }
+		} else {
+			startsOrEndsWithWhitespace = false;
+		}
+		return startsOrEndsWithWhitespace;		
+	}
+	
+	private boolean matchNamePattern(String value, String namePattern) {
+		boolean match = true;
+		if (value != null && StringUtils.isNotBlank(value)) {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: Doesn't StringUtils.isNotBlank already compare for null?

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 11]
```diff
@@ -164,4 +180,29 @@ private String getFieldKey(String field, boolean arrayInd, boolean testInd) {
 		return testInd ? field : arrayInd ? "names[0]." + field : "name." + field;
 	}
 	
-}
+	
+	private boolean startsOrEndsWithWhitespace(String value) {
+		boolean startsOrEndsWithWhitespace = false;
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: I think, here you made more complex for this function. Can we think like this way to reduce the nesting? 

```
private boolean startsOrEndsWithWhitespace(String value) {
   return (StringUtils.isNotBlank(value) && (Character.isSpaceChar(value.charAt(0)) || Character.isSpaceChar(value.charAt(-1)));
}
```

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java [hunk 12]
```diff
@@ -164,4 +180,29 @@ private String getFieldKey(String field, boolean arrayInd, boolean testInd) {
...
+			    startsOrEndsWithWhitespace = true; 
+		    }
+		} else {
+			startsOrEndsWithWhitespace = false;
+		}
+		return startsOrEndsWithWhitespace;		
+	}
+	
+	private boolean matchNamePattern(String value, String namePattern) {
+		boolean match = true;
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: Good to make this without the `if` condition. Just enough

`return StringUtils.isNotBlank(value) && value.matches(namePattern); `

File: api/src/main/java/org/openmrs/validator/PersonNameValidator.java (hunk 10 above)
Comment (suthagar23) [reply api/src/main/java/org/openmrs/validator/PersonNameValidator.java]: @dkayiwa  I think, the `StingUtils.isNotBlank` has the capability to check for `null` also. 
```
 StringUtils.isNotBlank(null)      = false
 StringUtils.isNotBlank("")        = false
 StringUtils.isNotBlank(" ")       = false
```

**Comment by dkayiwa** [comment]
@samuelmale did you see the comments on this?

**Comment by samuelmale** [comment]
Just addressed the comments. Thanks @suthagar23

**Comment by samuelmale** [comment]
@dkayiwa ping

**Comment by dkayiwa** [comment]
@samuelmale did you see the merge conflicts on this?

**Pull request closed without merging**